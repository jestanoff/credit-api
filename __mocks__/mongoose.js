export const mocks = {
  find: jest.fn(),
  findOne: jest.fn().mockReturnValue({ cardId: 'singleCard' }),
  limit: jest.fn().mockReturnValue([{ id: 1 }, { id: 2 }]),
  path: jest.fn(),
  save: jest.fn(),
};

const mock = {
  model: jest.fn().mockImplementation(() => ({
    find: mocks.find.mockImplementation(() => ({ limit: mocks.limit })),
    findOne: mocks.findOne,
    save: mocks.save,
  })),
  Schema: jest.fn().mockImplementation((definition, options) => ({
    definition,
    path: mocks.path,
    options,
  })),
};

export default mock;
