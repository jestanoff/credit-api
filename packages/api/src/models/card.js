import dynamoose from 'dynamoose';

dynamoose.aws.sdk.config.update({
  accessKeyId: process.env.AWS_ACCESS_KEY_ID,
  secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
  region: process.env.AWS_REGION,
});
// dynamoose.aws.ddb.local();

const CardSchema = new dynamoose.Schema(
  {
    id: {
      type: String,
      hashKey: true,
      required: true,
    },
    balance: {
      type: Number,
      default: 0,
      validate: val => val >= 0 && val <= 9999,
    },
    transactions: {
      type: Array,
      default: [],
      schema: [{
        type: Object,
        schema: {
          amount: {
            type: Number,
            validate: val => val >= 0 && val <= 9999,
            required: true,
          },
          createdAt: {
            type: Date,
            // TODO: there is a bug with nested schemas containing both required and default attr
            // work around is to require createdAt and then always set it to the current timestamp
            // default: Date.now,
            set: Date.now,
            required: true,
          },
          type: {
            type: String,
            enum: ['deposit', 'withdraw'],
            required: true,
          },
        },
      }],
    },
  },
  {
    timestamps: true,
  },
);
const Card = dynamoose.model('Cards', CardSchema);

export const getCards = () => Card.scan().exec();

export const getCard = id => Card.get({ id });

export const createCard = async id => {
  const existingCard = await getCard(id);
  if (existingCard && existingCard.id) {
    const err = new Error(`Card with id ${id} already exists`);
    err.code = 'CARD_ALREADY_EXISTS';
    return Promise.reject(err);
  }
  return new Card({ id }).save();
};

export const amendBalance = async (id, amount) => {
  if (!id) {
    const err = new Error('cardId is required');
    err.code = 'CARD_ID_REQUIRED';
    throw err;
  }

  if (amount === undefined) {
    const err = new Error('amount is required');
    err.code = 'AMOUNT_REQUIRED';
    throw err;
  }

  const transaction = {
    amount: Math.abs(amount),
    createdAt: Date.now(),
    type: Math.sign(amount) === -1 ? 'withdraw' : 'deposit',
  };

  // Using dynamoose.Condition to minimize querying the document
  // as checking for insufficient balance is needed on withdraw
  const conditionAmount = Math.sign(amount) === -1 ? Math.abs(amount) : 0;
  const condition = new dynamoose.Condition().filter('balance').ge(Math.abs(conditionAmount));
  return Card.update(
    { id },
    { $ADD: { balance: amount, transactions: transaction } },
    { condition },
  );
};
