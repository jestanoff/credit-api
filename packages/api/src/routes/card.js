import {
  amendBalance, createCard, getCard, getCards,
} from '../models/card.js';

const BAD_REQUEST = { error: 'Bad request' };

export const list = (req, res) => {
  try {
    const cards = getCards();
    return res.status(200).json(cards);
  } catch (err) {
    console.error(err);
    return res.status(500).json({ error: err.toString() }).end();
  }
};

export const create = (req, res) => {
  if (req && req.params && req.params.id) {
    return createCard(req.params.id)
      .then(card => res.status(201).json(card))
      .catch(err => {
        if (err.code && err.code === 'CARD_ALREADY_EXISTS') {
          return res
            .status(409)
            .json({ code: err.code, message: err.message });
        }
        console.log(err);
        return res.status(500).json({ error: err.toString() }).end();
      });
  }
  return res.status(400).json(BAD_REQUEST).end();
};

export const show = (req, res) => {
  if (req && req.params && req.params.id) {
    return getCard(req.params.id)
      .then(card => {
        if (!card) return res.status(404);
        return res.status(200).json(card);
      })
      .catch(err => {
        console.error(err);
        return res.status(500).json({ error: err.toString() }).end();
      });
  }
  return res.status(400).json(BAD_REQUEST).end();
};

export const balance = (req, res) => {
  if (req && req.params && req.params.id) {
    return getCard(req.params.id)
      .then(card => {
        if (card === undefined) return res.status(404);
        return res.status(200).json({ balance: card.balance });
      })
      .catch(err => {
        console.error(err);
        return res.status(500).json({ error: err.toString() }).end();
      });
  }
  return res.status(400).json(BAD_REQUEST).end();
};

export const deposit = async (req, res) => {
  if (req && req.params && req.params.id && req.body && req.body.amount && req.body.amount > 0) {
    try {
      const card = await amendBalance(req.params.id, req.body.amount);

      return res.status(200).json({ balance: card.balance });
    } catch (err) {
      if (err.code && err.code === 'ValidationException') {
        return res.status(404);
      }
      console.error(err);
      return res.status(500).json({ error: err.toString() }).end();
    }
  }
  return res.status(400).json(BAD_REQUEST).end();
};

export const withdraw = async (req, res) => {
  if (req && req.params && req.params.id && req.body && req.body.amount && req.body.amount > 0) {
    const { params: { id }, body: { amount } } = req;
    try {
      const updatedCard = await amendBalance(id, -amount);

      return res.status(200).json({ balance: updatedCard.balance });
    } catch (err) {
      if (err.code && err.code === 'ValidationException') {
        return res.status(404);
      }

      if (err.code && err.code === 'ConditionalCheckFailedException') {
        return res
          .status(409)
          .json({ code: 'INSUFFICIENT_BALANCE', message: 'Not enough credits on the card' });
      }

      console.error(err);
      return res.status(500).json({ error: err.toString() }).end();
    }
  }
  return res.status(400).json(BAD_REQUEST).end();
};
