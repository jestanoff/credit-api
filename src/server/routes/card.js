import {
  amendBalance, createCard, getCard, getCards,
} from '../models/card.js';

export const list = (req, res) => getCards()
  .then(cards => res.status(200).json(cards))
  .catch(console.error);

export const create = (req, res) => {
  if (req.params.id) {
    return createCard(req.params.id)
      .then(card => res.status(201).json(card))
      .catch(err => {
        if (err.name === 'ValidationError') {
          res
            .status(409)
            .json({ code: 'CARD_ALREADY_EXISTS', message: err.errors.cardId.message });
        } else {
          console.error(err);
        }
      });
  }
  return res.status(400).end();
};

export const show = (req, res) => {
  if (req.params.id) {
    return getCard(req.params.id)
      .then(card => res.status(200).json(card))
      .catch(console.error);
  }
  return res.status(400).end();
};

export const balance = (req, res) => {
  if (req.params.id) {
    return getCard(req.params.id)
      .then(card => res.status(200).json({ balance: card.balance }))
      .catch(console.error);
  }
  return res.status(400).end();
};

export const deposit = async (req, res) => {
  if (req.params.id && req.body.amount) {
    return amendBalance(req.params.id, req.body.amount)
      .then(card => res.status(200).json({ balance: card.balance }))
      .catch(console.error);
  }
  return res.status(400).end();
};

export const withdraw = async (req, res) => {
  if (req.params.id && req.body.amount) {
    try {
      const card = await amendBalance(req.params.id, -req.body.amount);

      if (card) return res.status(200).json({ balance: card.balance });

      return res
        .status(409)
        .json({ code: 'INSUFFICIENT_BALANCE', message: 'Not enough credits on the card' });
    } catch (err) {
      console.error(err);
    }
  }
  return res.status(400).end();
};
