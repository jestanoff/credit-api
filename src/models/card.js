import mongoose from 'mongoose';

const { Schema } = mongoose;
const TransactionSchema = new Schema(
  {
    amount: {
      max: 1000,
      min: 1,
      required: true,
      type: Number,
    },
    type: {
      required: true,
      type: String,
      enum: ['deposit', 'withdraw'],
    },
  },
  {
    autoIndex: false,
    strict: true,
    timestamps: true,
  },
);
const CardSchema = new Schema(
  {
    balance: {
      default: 0,
      max: 1000,
      min: 0,
      type: Number,
    },
    transactions: {
      type: [TransactionSchema],
    },
  },
  {
    autoIndex: false,
    strict: true,
    timestamps: true,
  },
);
const Card = mongoose.model('Card', CardSchema);
const Transaction = mongoose.model('Transaction', TransactionSchema);

export const getCards = (limit = 50) => Card.find().limit(limit);

export const createCard = () => new Card({}).save();

export const getCard = id => Card.findById(id);

export const amendBalance = async (id, amount) => {
  const card = await Card.findById(id);
  const balance = card.balance + amount;

  if (balance < 0) return undefined;

  card.balance = balance;
  card.transactions.push(new Transaction({
    amount: Math.abs(amount),
    type: Math.sign(amount) === -1 ? 'withdraw' : 'deposit',
  }));
  card.save();
  return card;
};
