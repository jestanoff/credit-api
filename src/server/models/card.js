import mongoose from 'mongoose';

const { Schema } = mongoose;
const TransactionSchema = new Schema(
  {
    amount: {
      max: 1000,
      min: 1,
      required: true,
      type: Number,
    },
    type: {
      enum: ['deposit', 'withdraw'],
      required: true,
      type: String,
    },
  },
  {
    autoIndex: false,
    strict: true,
    timestamps: true,
  },
);
const CardSchema = new Schema(
  {
    balance: {
      default: 0,
      max: 1000,
      min: 0,
      type: Number,
    },
    cardId: String,
    transactions: {
      type: [TransactionSchema],
    },
  },
  {
    autoIndex: false,
    strict: true,
    timestamps: true,
  },
);
const Card = mongoose.model('Card', CardSchema);
CardSchema.path('cardId', {
  required: true,
  type: String,
  unique: true,
  validate: {
    validator: async (value) => {
      const card = await Card.findOne({ cardId: value });

      return card
        ? Promise.reject(new Error(`Card with id ${card.cardId} already exists`))
        : Promise.resolve('Card created');
    },
  },
});

const Transaction = mongoose.model('Transaction', TransactionSchema);

export const getCards = (limit = 50) => Card.find().limit(limit);

export const createCard = cardId => new Card({ cardId }).save();

export const getCard = cardId => Card.findOne({ cardId });

export const amendBalance = async (cardId, amount) => {
  const card = await Card.findOne({ cardId });
  const balance = card.balance + amount;

  if (balance < 0) return undefined;

  const transactions = [...card.transactions,
    new Transaction({
      amount: Math.abs(amount),
      type: Math.sign(amount) === -1 ? 'withdraw' : 'deposit',
    }),
  ];
  return Card.findOneAndUpdate({ cardId }, { balance, transactions }, { new: true });
};
