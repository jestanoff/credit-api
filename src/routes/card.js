import { createCard, getCard, getCards } from '../models/card.js';

export const list = (req, res) => {
  getCards().then(cards => {
    res.status(200).json(cards);
  }).catch(console.error);
};

export const create = (req, res) => {
  createCard().then(card => {
    res.status(401).json(card);
  }).catch(console.error);
};

export const show = (req, res) => {
  if (req.params.id) {
    getCard(req.params.id).then(card => {
      res.status(200).json(card[0]);
    }).catch(console.error);
  } else {
    res.status(400).end();
  }
};

export const balance = (req, res) => {
  res.end(`The balance of card #${req.params.id} is ${req.params.id * 1.33} credits`);
};

export const deposit = (req, res) => {
  res.end(`Deposited to card #${req.params.id}, ${req.params.id * 0.63} credits`);
};

export const withdraw = (req, res) => {
  res.end(`Withdrew from card #${req.params.id}, ${req.params.id * 0.77} credits`);
};
