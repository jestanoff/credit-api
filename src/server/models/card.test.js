import mongoose, { mocks } from 'mongoose';
// eslint-disable-next-line no-unused-vars
import * as card from './card.js';

jest.mock('mongoose');

describe('Card model', () => {
  test('should create a Card schema', () => {
    expect(mongoose.Schema.mock.calls[1]).toMatchSnapshot();
  });

  test('should create a new car document with Card schema', () => {
    expect(mongoose.model).toBeCalled();
    expect(mongoose.model.mock.calls[0]).toMatchSnapshot();
  });

  test('should path the Card schema with a carId validator', () => {
    expect(mocks.path).toBeCalledTimes(1);
    expect(mocks.path).toBeCalledWith('cardId', {
      required: true,
      type: String,
      unique: true,
      validate: {
        validator: expect.any(Function),
      },
    });
  });

  describe('validation on card creation to ensure cardId uniqueness', () => {
    beforeEach(() => mocks.findOne.mockClear());

    test('should reject the promise if it finds a card', () => {
      expect.assertions(3);
      const error = new Error('Card with id singleCard already exists');
      const { validator } = mocks.path.mock.calls[0][1].validate;

      validator('cardId').catch(err => {
        expect(err).toEqual(error);
        expect(mocks.findOne).toBeCalledTimes(1);
        expect(mocks.findOne).toBeCalledWith({ cardId: 'cardId' });
      });
    });

    test("should resolve the promise if it doesn't find a card", async () => {
      expect.assertions(3);
      mocks.findOne.mockReturnValue(undefined);
      const { validator } = mocks.path.mock.calls[0][1].validate;
      const promise = await validator('Hello');

      expect(promise).toBe('Card created');
      expect(mocks.findOne).toBeCalledTimes(1);
      expect(mocks.findOne).toBeCalledWith({ cardId: 'Hello' });
    });
  });
});

describe('Transaction model', () => {
  test('should create a Transaction schema', () => {
    expect(mongoose.Schema).toBeCalledTimes(2);
    expect(mongoose.Schema.mock.calls[0]).toMatchSnapshot();
  });

  test('should create a new transaction document with Transaction schema', () => {
    expect(mongoose.model).toBeCalled();
    expect(mongoose.model.mock.calls[1]).toMatchSnapshot();
  });
});

describe('db queries', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mocks.findOne.mockReturnValue({
      balance: 20,
      cardId: 'singleCard',
      transactions: ['1', '2'],
    });
  });

  describe('getCards', () => {
    test('should return up to 50 cards ', () => {
      expect(card.getCards()).toMatchObject([{ id: 1 }, { id: 2 }]);
      expect(mocks.find).toBeCalledTimes(1);
      expect(mocks.find).toBeCalledWith();
      expect(mocks.limit).toBeCalledTimes(1);
      expect(mocks.limit).toBeCalledWith(50);
    });

    test('should return up to 5 cards when limit is set to 5', () => {
      card.getCards(5);

      expect(mocks.limit).toBeCalledWith(5);
    });
  });

  describe('createCard', () => {
    test('should crate a new card with a given cardId and then save it', () => {
      // TODO
      // card.createCard('newCard');
    });
  });

  describe('getCard', () => {
    test('should get a single card with a given cardId', () => {
      expect(card.getCard('getMeThatCard')).toMatchObject({
        balance: 20,
        cardId: 'singleCard',
        transactions: ['1', '2'],
      });
      expect(mocks.findOne).toBeCalledTimes(1);
      expect(mocks.findOne).toBeCalledWith({ cardId: 'getMeThatCard' });
    });
  });

  describe('amendBalance', () => {
    test('should only amend the balance of a single card with a given cardId', async () => {
      // TODO
      // await card.amendBalance('getMeThisCard', 10);

      // expect(mocks.findOne).toBeCalledTimes(1);
      // expect(mocks.findOne).toBeCalledWith({ cardId: 'getMeThisCard' });
    });
  });
});
