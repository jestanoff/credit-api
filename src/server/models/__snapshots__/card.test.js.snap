// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card model should create a Card schema 1`] = `
Array [
  Object {
    "balance": Object {
      "default": 0,
      "max": 1000,
      "min": 0,
      "type": [Function],
    },
    "cardId": [Function],
    "transactions": Object {
      "type": Array [
        Object {
          "definition": Object {
            "amount": Object {
              "max": 1000,
              "min": 1,
              "required": true,
              "type": [Function],
            },
            "type": Object {
              "enum": Array [
                "deposit",
                "withdraw",
              ],
              "required": true,
              "type": [Function],
            },
          },
          "options": Object {
            "autoIndex": false,
            "strict": true,
            "timestamps": true,
          },
          "path": [MockFunction] {
            "calls": Array [
              Array [
                "cardId",
                Object {
                  "required": true,
                  "type": [Function],
                  "unique": true,
                  "validate": Object {
                    "validator": [Function],
                  },
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
        },
      ],
    },
  },
  Object {
    "autoIndex": false,
    "strict": true,
    "timestamps": true,
  },
]
`;

exports[`Card model should create a new car document with Card schema 1`] = `
Array [
  "Card",
  Object {
    "definition": Object {
      "balance": Object {
        "default": 0,
        "max": 1000,
        "min": 0,
        "type": [Function],
      },
      "cardId": [Function],
      "transactions": Object {
        "type": Array [
          Object {
            "definition": Object {
              "amount": Object {
                "max": 1000,
                "min": 1,
                "required": true,
                "type": [Function],
              },
              "type": Object {
                "enum": Array [
                  "deposit",
                  "withdraw",
                ],
                "required": true,
                "type": [Function],
              },
            },
            "options": Object {
              "autoIndex": false,
              "strict": true,
              "timestamps": true,
            },
            "path": [MockFunction] {
              "calls": Array [
                Array [
                  "cardId",
                  Object {
                    "required": true,
                    "type": [Function],
                    "unique": true,
                    "validate": Object {
                      "validator": [Function],
                    },
                  },
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
          },
        ],
      },
    },
    "options": Object {
      "autoIndex": false,
      "strict": true,
      "timestamps": true,
    },
    "path": [MockFunction] {
      "calls": Array [
        Array [
          "cardId",
          Object {
            "required": true,
            "type": [Function],
            "unique": true,
            "validate": Object {
              "validator": [Function],
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
]
`;

exports[`Transaction model should create a Transaction schema 1`] = `
Array [
  Object {
    "amount": Object {
      "max": 1000,
      "min": 1,
      "required": true,
      "type": [Function],
    },
    "type": Object {
      "enum": Array [
        "deposit",
        "withdraw",
      ],
      "required": true,
      "type": [Function],
    },
  },
  Object {
    "autoIndex": false,
    "strict": true,
    "timestamps": true,
  },
]
`;

exports[`Transaction model should create a new transaction document with Transaction schema 1`] = `
Array [
  "Transaction",
  Object {
    "definition": Object {
      "amount": Object {
        "max": 1000,
        "min": 1,
        "required": true,
        "type": [Function],
      },
      "type": Object {
        "enum": Array [
          "deposit",
          "withdraw",
        ],
        "required": true,
        "type": [Function],
      },
    },
    "options": Object {
      "autoIndex": false,
      "strict": true,
      "timestamps": true,
    },
    "path": [MockFunction] {
      "calls": Array [
        Array [
          "cardId",
          Object {
            "required": true,
            "type": [Function],
            "unique": true,
            "validate": Object {
              "validator": [Function],
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
]
`;
